File: backend/src/utils/roomCode.js
// Deliberately excludes visually ambiguous characters (0/O, 1/I).
const ALPHABET = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';

function randomCode(length = 5) {
  let out = '';
  for (let i = 0; i < length; i += 1) {
    out += ALPHABET[Math.floor(Math.random() * ALPHABET.length)];
  }
  return out;
}

function generateUniqueCode(existingCodes, length = 5) {
  let code = randomCode(length);
  let attempts = 0;
  while (existingCodes.has(code) && attempts < 50) {
    code = randomCode(length);
    attempts += 1;
  }
  return code;
}

module.exports = { generateUniqueCode };

File: backend/src/room_code.py
"""Short, human-friendly room codes."""

import random

# Deliberately excludes visually ambiguous characters (0/O, 1/I).
ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"


def _random_code(length: int = 5) -> str:
    return "".join(random.choice(ALPHABET) for _ in range(length))


def generate_unique_code(existing_codes: set[str], length: int = 5) -> str:
    code = _random_code(length)
    attempts = 0
    while code in existing_codes and attempts < 50:
        code = _random_code(length)
        attempts += 1
    return code
